### PR TITLE
#14001: De-globalize reflect

### PR DESCRIPTION
**Pull request by afuller-TT**
### Ticket
#14001 

### Problem description
Fixing globals in service of being able to package up Metal.

### What's changed
Object-ify `reflect` and consume it where it's actually used, allowing it to propagate as needed.
Remove the global reference to `reflect`.

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/11632757225
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes
- [x] TTNN C++ Project passes https://github.com/tenstorrent/tt-metal/actions/runs/11618754311

### CODE REVIEW

**Comment by tt-rkim** [comment]
Does moving this behind `reflect_ADDED` and using it only where it's needed change how we need to package? Can't reflect still be used in metal's headers?

**Comment by afuller-TT** [comment]
> Does moving this behind `reflect_ADDED` and using it only where it's needed change how we need to package? Can't reflect still be used in metal's headers?

It absolutely could.  But this gives more information to CMake.  If we declare `reflect` as a PUBLIC dependency, then when packaging CMake will tell us "You can't package Metal, unless you include reflect". vs if `reflect` is a PRIVATE dependency, then CMake understands that we can package the Metal binary without needing to also package reflect.

Current state is that the path is just added globally and neither CMake nor us developers know where it's actually used and whether it needs to be included in our package or not.

File: tt_metal/CMakeLists.txt [hunk 1]
```diff
@@ -26,6 +26,8 @@ target_link_libraries(
     tt_metal
     PUBLIC
         metal_header_directories
+        impl
```
Comment (blozano-tt) [inline tt_metal/CMakeLists.txt:29]: why are you not using the alias you created?

File: tt_metal/CMakeLists.txt [hunk 2]
```diff
@@ -26,6 +26,8 @@ target_link_libraries(
...
+        Metalium::Metal::stl
```
Comment (blozano-tt) [inline tt_metal/CMakeLists.txt:30]: Does it have to be public?

**Review by blozano-tt — APPROVED** [review]
Thanks for making the world a better place.

**Review by tt-rkim — APPROVED** [review]

**Review by TT-billteng — APPROVED** [review]

**Pull request merged**